src/coreclr/vm/threadstatics.cpp

```diff
@@ -799,6 +803,34 @@ bool CanJITOptimizeTLSAccess()
     // Optimization is disabled for FreeBSD/arm64
 #elif defined(FEATURE_INTERPRETER)
     // Optimization is disabled when interpreter may be used
+#elif !defined(TARGET_OSX) && defined(TARGET_UNIX) && defined(TARGET_ARM64)
+    // Optimization is enabled for linux/arm64 only for static resolver.
+    // For static resolver, the TP offset is same for all threads.
+    // For dynamic resolver, TP offset returned is for the current thread and
+    // will be different for the other threads.
+    uint32_t* resolverAddress = reinterpret_cast<uint32_t*>(GetTLSResolverAddress());
+    int ip = 0;
+    if ((resolverAddress[ip] == 0xd503201f) || (resolverAddress[ip] == 0xd503241f))
+    {
+        // nop might not be present in older resolver, so skip it.
+
+        // nop or hint 32
+        ip++;
+    }
+
+    if (
+        // ldr x0, [x0, #8]
+        (resolverAddress[ip] == 0xf9400400) &&
+        // ret
+        (resolverAddress[ip + 1] == 0xd65f03c0)
+    )
+    {
+        optimizeThreadStaticAccess = true;
+    }
+    else
+    {
+        _ASSERTE(false && "Unexpected code sequence.");
```

Comment:
Can we add a test that validates this fallback? I expect this question is going to be asked during servicing review.

Reply:
Make sense. What is the right way to add a test? Force dynamic resolver to kick in or something? Need to find out how to do that though.

Reply:
Rigth:

- Add a native test library with a lot of thread statics (look for CMakeLists.txt under src\tests for how to add one of those). The thread statics need to be individual thread static variables so that they are allocated individually. Loading this library should eat all non-dynamic statics space and force the dynamic resolver to kick in.

- Create a test that loads this native test library before loading libcoreclr.so. I am not sure what's the best way to do that. You can try using `LD_PRELOAD` that points to the native test library; or you can try to load it as a mock host policy using [this hook](https://github.com/dotnet/runtime/blob/9efc798f850e2bb68c02382c393d2cb1e818eb45/src/coreclr/hosts/corerun/corerun.cpp#L298-L306) in corerun. cc @jkoritzinsky and @AaronRobinsonMSFT for more thoughts.

Reply:
> Add a native test library with a lot of thread statics 

Interesting, my impression was using some kind of compiler flag or an attribute. 

> Loading this library should eat all non-dynamic statics space and force the dynamic resolver to kick in. 

But even if I get it to use the dynamic resolver, the question would be how to check that the TLS optimization is disabled I suppose. I believe I will have to still write managed code containing `ThreadStatic` variable and I will have to verify that its access was not optimized. That will need somehow getting hold of the jitted method's address and reading the instruction stream of it to confirm that we did not do TLS access? Alternatively, just add a check here that the instruction stream we get is that of the dynamic resolver, but not sure if we will test anything interesting with that.

Reply:
Instead of this assert, you can add a debug-only config switch that says whether we expect dynamic or static resolver. If we come here and we do not the expected resolver, assert. I think it is good-enough for the test to be effective on checked build only.

Reply:
> just add a check here that the instruction stream we get is that of the dynamic resolver, but not sure if we will test anything interesting with that.

The test should do some thread static accesses from multiple threads to make sure that they work.

Reply:
> But even if I get it to use the dynamic resolver, the question would be how to check that the TLS optimization is disabled I suppose. I believe I will have to still write managed code containing `ThreadStatic` variable and I will have to verify that its access was not optimized. That will need somehow getting hold of the jitted method's address and reading the instruction stream of it to confirm that we did not do TLS access? Alternatively, just add a check here that the instruction stream we get is that of the dynamic resolver, but not sure if we will test anything interesting with that.

You can use [disasm checks](https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/disasm-checks.md) to do this too.

Reply:
> Create a test that loads this native test library before loading libcoreclr.so. I am not sure what's the best way to do that. You can try using LD_PRELOAD that points to the native test library; or you can try to load it as a mock host policy using this hook in corerun. cc @jkoritzinsky and @AaronRobinsonMSFT for more thoughts.

Seems like a reasonable feature. I would create a new environment variables that loads an arbitrary DLL early on. Once tht is in, I will update the host policy mechanism to use that.